### PR TITLE
Improved Flatpak and Snapcraft packaing

### LINKED ISSUE #1435: Must upload stem fork to PyPi
I'm working on making the version 2.4 release and I just hit a snag with publishing to PyPi:

```
$ poetry publish --build
There are 2 files ready for publishing. Build anyway? (yes/no) [no] yes
Building onionshare_cli (2.4)
  - Building sdist
  - Built onionshare_cli-2.4.tar.gz
  - Building wheel
  - Built onionshare_cli-2.4-py3-none-any.whl

Username: micahflee
Password: 
Publishing onionshare_cli (2.4) to PyPI
 - Uploading onionshare_cli-2.4-py3-none-any.whl 100%

  UploadError

  HTTP Error 400: Invalid value for requires_dist. Error: Can't have direct dependency: 'stem @ git+https://github.com/onionshare/stem.git@1.8.1'

  at ~/.local/lib/python3.9/site-packages/poetry/publishing/uploader.py:216 in _upload
      212│                     self._register(session, url)
      213│                 except HTTPError as e:
      214│                     raise UploadError(e)
      215│ 
    → 216│             raise UploadError(e)
      217│ 
      218│     def _do_upload(
      219│         self, session, url, dry_run=False
      220│     ):  # type: (requests.Session, str, Optional[bool]) -> None
```

PyPi doesn't support git repo dependencies, so we can't publish onionshare 2.4 there.

For this to work, we'll need to upload stem 1.8.1 to PyPi, but the name `stem` is reserved for the original project, and this is a fork. So I guess we'll need to rename it?

What are your thoughts @mig5 @SaptakS ?

### LINKED ISSUE #1437: OnionShare fails to start due to Apparmor DENIED in Ubuntu 21.04
Version: 2.4, installed via snap
OS: Ubuntu 20.04.3

Typical Behavior: OnionShare launches normally (as of previous recent version, uncertain of exactly which. Last used OnionShare successfully around 2 or 3 months ago).

Current Behavior (as of v2.4): When lauching OnionShare from GUI, process launches, hangs, then shuts down without displaying a window. When attempting to launch from CLI, the following error message is displayed:

```
qt.qpa.xcb: could not connect to display :0
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, webgl, xcb.

Aborted (core dumped)
```

Attempting to uninstall/reinstall, did not affect issue. Attempted system reboot to shake things loose, but also did not affect issue. OnionShare currently does not function.

I know of one other person who had a similar issue (which led me to try reproducing it). I was unable to find out which OS they were using at the time, but I believe it to be Windows based on screenshots. Their issue was similar to mine (unable to launch program) but the content of their error message was slightly different:

```
pythonw

This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem
```

### LINKED ISSUE #1459: Make snap and flatpak build snowflake-client and meek-client binaries from source

### LINKED ISSUE #1466: Onionshare Doesn't Open On Ubuntu 20.04
I used snap to install. When I clicked on the icon nothing happened so I went to the command line:

```
$ onionshare
Traceback (most recent call last):
  File "/snap/onionshare/14/bin/onionshare", line 33, in <module>
    sys.exit(load_entry_point('onionshare==2.4', 'console_scripts', 'onionshare')())
  File "/snap/onionshare/14/lib/python3.6/site-packages/onionshare/__init__.py", line 137, in main
    common.display_banner()
  File "/snap/onionshare/14/lib/python3.6/site-packages/onionshare_cli/common.py", line 95, in display_banner
    Back.MAGENTA + Fore.WHITE + "\u256d\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u256e"
  File "/snap/onionshare/14/lib/python3.6/site-packages/colorama/ansitowin32.py", line 41, in write
    self.__convertor.write(text)
  File "/snap/onionshare/14/lib/python3.6/site-packages/colorama/ansitowin32.py", line 164, in write
    self.wrapped.write(text)
UnicodeEncodeError: 'ascii' codec can't encode characters in position 10-54: ordinal not in range(128)
```

### PR DESCRIPTION
**Pull request by micahflee**
This PR, along with the `onionshare-2.4.1` branch of the [Flatpak repository](https://github.com/flathub/org.onionshare.OnionShare/tree/onionshare-2.4.1), fixes #1459.

In both Flatpak and Snapcraft, `snowflake-client` and `meek-client` are built from source, and I made some modifications to the code so that it finds them correctly

This also increases the security of the Flatpak package by revoking the `--filesystem=home:ro` permission that gives OnionShare read-only permission to the user's home dir. Instead, OnionShare only has permission to read and write to its config folder, as well as `~/OnionShare`. In share and website modes, there are now two separate buttons, "Add Files" and "Add Folder", and these use Flatpak portals to basically give access to files that are browsed for using those file selection dialogs.

Unfortunately, Flatpak doesn't support drag-and-drop portals, so if you're running the Flatpak package, I've disabled drag-and-drop. If you try to drag files in, it shows you this popup telling you to use the buttons instead:

Screenshot from 2021-11-15 18-48-58

This also fixes #1435 by switching from stem to [cepa](https://github.com/onionshare/cepa) 1.8.3, which is nice because it simplifies the Flatpak packaging some.

I haven't tested it because I haven't reproduced it, but this _should_ fix #1466 as well. In some OSes it looks like OnionShare crashes just trying to display the pretty ASCII art, so this commit https://github.com/onionshare/onionshare/commit/267e9eb1e43bb188e8925caecb71914130fb266f just does exception handling on the OnionShare banner, and if it hits an exception it skips the ASCII art and displays a much simpler banner.

This also _should_ fix #1437, but it's hard to be completely certain before the new snap is pushed in the Snap Store. But I was able to build a local snap, install it locally like `snap install --dangerous onionshare_2.4.1_amd64.snap`, which installs it with the sandbox enabled, and I'm not reproducing the crash. (I possibly fixed it by adding the `gnome-3-34` extension to the snap.)

## To test Flatpak

Get the code:

```
git clone https://github.com/flathub/org.onionshare.OnionShare
cd org.onionshare.OnionShare
git checkout onionshare-2.4.1
```

Build and install the Flatpak package, which takes awhile. You need `flatpak` and `flatpak-builder` to be installed.

```
flatpak-builder build --force-clean --install-deps-from=flathub --install --user org.onionshare.OnionShare.yaml
```

Then test it:

```
flatpak run org.onionshare.OnionShare
```

## To test Snapcraft

Make sure you're in the onionshare repo directory, and you have `snap` installed and also `snapcraft` installed (`snap install snapcraft --classic`).

Build the OnionShare snap:

```
snapcraft
```

This takes awhile. When it's done, install it:

```
sudo snap install --dangerous ./onionshare_2.4.1_amd64.snap
```

And run it:

```
/snap/bin/onionshare -v
```

### CODE REVIEW

File: cli/onionshare_cli/common.py [hunk 1]
```diff
@@ -432,6 +441,18 @@ def build_username(self, word_count=2):
...
 
+    def is_flatpak(self):
+        """
+        Returns True if OnionShare is running in a Flatpak sandbox
+        """
+        return os.environ.get("FLATPAK_ID") == "org.onionshare.OnionShare"
+
+    def is_snapcraft(self):
+        """
+        Returns True if OnionShare is running in a Flatpak sandbox
```
Comment (mig5) [inline cli/onionshare_cli/common.py:452]: nitpick: 'is running in a *Snapcraft* sandbox'

**Review by mig5 — COMMENTED** [review]
This is looking mostly good so far! I have only tested the Snap, my Flatpak is still building. I'll update again once I've tested it.

Found one tiny nitpick in a code comment.

I also observed, with the Snap, that after downloading a share (which was in the default 'stop after first download' mode), the History pane shows an 'in progress' progress bar still spinning, even though the download succeeded, and the share has definitely stopped.

I don't reproduce this in native Linux on this branch, so it seems unique to Snap at least so far. It's not really a major thing, just a curiosity, and it might predate this work, I'm not sure. Could it mean that an event is not being received from the backend to the frontend to update the widget?

Screenshot_2021-11-16_16-41-25

**Comment by mig5** [comment]
I tried Flatpak today. There are a couple of problems.

1. Clicking 'Add Files' doesn't do anything. No apparent stack trace or anything according to verbose mode. Just simply doesn't do anything. So at this point, I can't share files unless they're in folders.
2. If you click 'Add Folder', and then click 'Cancel' or otherwise close the File selection dialog, you get an Alert dialog ' is not a readable file'.

Screenshot_2021-11-17_07-48-27

3. I reproduce the 'spinning progress bar' HistoryItem when a share stops and the share had been fully downloaded by the client, which I saw in Snap but not in native Linux (via `scripts/dev.sh`)

But sharing a folder of files is working fine. The BridgeDB/meek system is working fine. I can't seem to start shared using the built-in Snowflake, but I don't get errors, so it might just be common intermittent timeout stuff happening with that bridge..

**Comment by micahflee** [comment]
It's really strange, but I haven't been able to reproduce either the History pane "in progress" thing, or the "Add Files" not working, in either Snapcraft or Flatpak... They both just work like they're supposed to.

I fixed the problem with clicking Add Folder then clicking cancel, and I also added logging to the `FileSelection.add_files` and `FileSelection.add_folder` so if the Add Files button still isn't working for you maybe we can at least troubleshoot some more. Here's what it should look like (in this case, I'm running the Flatpak version).

One a new share mode tab:

```
[Nov 17 2021 07:53:22 PM] Tab.share_mode_clicked
[Nov 17 2021 07:53:22 PM] Web.__init__: is_gui=True, mode=share
[Nov 17 2021 07:53:22 PM] Common.get_resource_path: filename=static
[Nov 17 2021 07:53:22 PM] Common.get_resource_path: filename=static, path=/home/user/.var/app/org.onionshare.OnionShare/cache/Python-Eggs/onionshare_cli-2.4-py3.8.egg-tmp/onionshare_cli/resources/static
[Nov 17 2021 07:53:22 PM] Common.get_resource_path: filename=templates
[Nov 17 2021 07:53:22 PM] Common.get_resource_path: filename=templates, path=/home/user/.var/app/org.onionshare.OnionShare/cache/Python-Eggs/onionshare_cli-2.4-py3.8.egg-tmp/onionshare_cli/resources/templates
[Nov 17 2021 07:53:22 PM] Web.generate_static_url_path: new static_url_path is /static_cgj5hcbbb7crfjaw3g5j4y246e
[Nov 17 2021 07:53:22 PM] ShareModeWeb.init
[Nov 17 2021 07:53:22 PM] ModeSettings.set: updating confining-upcoming-ultimate: share.filenames = []
[Nov 17 2021 07:53:22 PM] ServerStatus.update
[Nov 17 2021 07:53:22 PM] ServerStatus.update
[Nov 17 2021 07:53:22 PM] ShareMode.update_primary_action
```

Click Add Files, browse, and select a bunch of screenshots (huh, it's cool that the paths of the screenshots are all in `/run/user/1000/doc/[hex]/[filename]`, this is the sandbox portal in action):

```
[Nov 17 2021 07:54:10 PM] FileSelection.add_files: (['/run/user/1000/doc/7f4e04ce/Screenshot from 2021-11-07 15-40-15.png', '/run/user/1000/doc/fe5c8633/Screenshot from 2021-11-08 07-17-10.png', '/run/user/1000/doc/479f111e/Screenshot from 2021-11-11 15-39-28.png', '/run/user/1000/doc/f33bcac3/Screenshot from 2021-11-12 17-28-39.png', '/run/user/1000/doc/6b80c6b3/Screenshot from 2021-11-14 14-07-04.png', '/run/user/1000/doc/4667ba04/Screenshot from 2021-11-15 18-48-58.png'], 'All Files (*)')
[Nov 17 2021 07:54:10 PM] ModeSettings.set: updating confining-upcoming-ultimate: share.filenames = ['/run/user/1000/doc/7f4e04ce/Screenshot from 2021-11-07 15-40-15.png']
[Nov 17 2021 07:54:10 PM] ServerStatus.update
[Nov 17 2021 07:54:10 PM] ShareMode.update_primary_action
[Nov 17 2021 07:54:10 PM] ModeSettings.set: updating confining-upcoming-ultimate: share.filenames = ['/run/user/1000/doc/7f4e04ce/Screenshot from 2021-11-07 15-40-15.png', '/run/user/1000/doc/fe5c8633/Screenshot from 2021-11-08 07-17-10.png']
[Nov 17 2021 07:54:10 PM] ServerStatus.update
[Nov 17 2021 07:54:10 PM] ShareMode.update_primary_action
[Nov 17 2021 07:54:10 PM] ModeSettings.set: updating confining-upcoming-ultimate: share.filenames = ['/run/user/1000/doc/7f4e04ce/Screenshot from 2021-11-07 15-40-15.png', '/run/user/1000/doc/fe5c8633/Screenshot from 2021-11-08 07-17-10.png', '/run/user/1000/doc/479f111e/Screenshot from 2021-11-11 15-39-28.png']
[Nov 17 2021 07:54:10 PM] ServerStatus.update
[Nov 17 2021 07:54:10 PM] ShareMode.update_primary_action
[Nov 17 2021 07:54:10 PM] ModeSettings.set: updating confining-upcoming-ultimate: share.filenames = ['/run/user/1000/doc/7f4e04ce/Screenshot from 2021-11-07 15-40-15.png', '/run/user/1000/doc/fe5c8633/Screenshot from 2021-11-08 07-17-10.png', '/run/user/1000/doc/479f111e/Screenshot from 2021-11-11 15-39-28.png', '/run/user/1000/doc/f33bcac3/Screenshot from 2021-11-12 17-28-39.png']
[Nov 17 2021 07:54:10 PM] ServerStatus.update
[Nov 17 2021 07:54:10 PM] ShareMode.update_primary_action
[Nov 17 2021 07:54:10 PM] ModeSettings.set: updating confining-upcoming-ultimate: share.filenames = ['/run/user/1000/doc/7f4e04ce/Screenshot from 2021-11-07 15-40-15.png', '/run/user/1000/doc/fe5c8633/Screenshot from 2021-11-08 07-17-10.png', '/run/user/1000/doc/479f111e/Screenshot from 2021-11-11 15-39-28.png', '/run/user/1000/doc/f33bcac3/Screenshot from 2021-11-12 17-28-39.png', '/run/user/1000/doc/6b80c6b3/Screenshot from 2021-11-14 14-07-04.png']
[Nov 17 2021 07:54:10 PM] ServerStatus.update
[Nov 17 2021 07:54:10 PM] ShareMode.update_primary_action
[Nov 17 2021 07:54:10 PM] ModeSettings.set: updating confining-upcoming-ultimate: share.filenames = ['/run/user/1000/doc/7f4e04ce/Screenshot from 2021-11-07 15-40-15.png', '/run/user/1000/doc/fe5c8633/Screenshot from 2021-11-08 07-17-10.png', '/run/user/1000/doc/479f111e/Screenshot from 2021-11-11 15-39-28.png', '/run/user/1000/doc/f33bcac3/Screenshot from 2021-11-12 17-28-39.png', '/run/user/1000/doc/6b80c6b3/Screenshot from 2021-11-14 14-07-04.png', '/run/user/1000/doc/4667ba04/Screenshot from 2021-11-15 18-48-58.png']
[Nov 17 2021 07:54:10 PM] ServerStatus.update
[Nov 17 2021 07:54:10 PM] ShareMode.update_primary_action
[Nov 17 2021 07:54:10 PM] ModeSettings.set: updating confining-upcoming-ultimate: share.filenames = ['/run/user/1000/doc/7f4e04ce/Screenshot from 2021-11-07 15-40-15.png', '/run/user/1000/doc/fe5c8633/Screenshot from 2021-11-08 07-17-10.png', '/run/user/1000/doc/479f111e/Screenshot from 2021-11-11 15-39-28.png', '/run/user/1000/doc/f33bcac3/Screenshot from 2021-11-12 17-28-39.png', '/run/user/1000/doc/6b80c6b3/Screenshot from 2021-11-14 14-07-04.png', '/run/user/1000/doc/4667ba04/Screenshot from 2021-11-15 18-48-58.png']
```

Then if I start the server and download it, this is my history pane:

Screenshot from 2021-11-17 19-57-08

**Comment by flipnut** [comment]
As far as I can see the read-only access permissions to the home folder have not been removed from the [Flatpak config](https://github.com/flathub/org.onionshare.OnionShare/blob/master/org.onionshare.OnionShare.yaml#L18) or the [Snapcraft config](https://github.com/onionshare/onionshare/blob/develop/snap/snapcraft.yaml#L21).

**Comment by micahflee** [comment]
@flipnut they have been removed from the flatpak config in the `onionshare-2.4.1` branch: https://github.com/flathub/org.onionshare.OnionShare/blob/onionshare-2.4.1/org.onionshare.OnionShare.yaml

I'm going to merge that into master when we do the actual 2.4.1 release.

Snapcraft still has home folder access though, because it seems that there's no way around that.

**Pull request merged**